File: problems/go/0318.go
func maxProduct(words []string) int {
	size := len(words)
	base := 'a'

	bitmap := [1001]int{0}
	for i, word := range words {
		for _, char := range word {
			pos := char - base
			bitmap[i] |= 1 << pos
		}
	}

	max := 0
	for i := 0; i < size; i++ {
		for j := i + 1; j < size; j++ {
			if (bitmap[i]&bitmap[j]) == 0 && len(words[i])*len(words[j]) > max {
				max = len(words[i]) * len(words[j])
			}
		}
	}

	return max
}

File: problems/js/0318.js
/**
 * @param {string[]} words
 * @return {number}
 */

// 定義 bitmap 物件
// length 為原始字串長度, value 為 bitmap 值
class bm {
    constructor(string) {
        this.length = string.length;
        this.set(string);
    }

    set(string) {
        let v = 0;
        Array.from(new Set(string.split(""))).forEach((char) => {
            // 97 為 "a" 的 char code
            v += Math.pow(2, char.charCodeAt(0) - 97);
        });
        this.value = v;
    }
}

const maxProduct = function (words) {
    // 暫存最大值
    let max = 0;
    // 把 words 從 string 轉成上面定義的 bm
    words = words.map((word) => new bm(word));

    // 一一配對
    for (let i = 0; i < words.length; i++) {
        for (let j = i + 1; j < words.length; j++) {
            // 如果兩個 bm 做 bitwise AND 每項皆為 0 ，代表無碰撞，即為合法組，試圖更新暫存最大值
            if (!(words[i].value & words[j].value)) max = words[i].length * words[j].length > max ? words[i].length * words[j].length : max;
        }
    }
    return max;
};
